crates/turbopack-ecmascript/src/tree_shake/mod.rs

```diff
@@ -0,0 +1,719 @@
+    #[turbo_tasks::function]
+    fn content(&self) -> AssetContentVc {
+        todo!()
+    }
+
+    #[turbo_tasks::function]
+    fn references(&self) -> AssetReferencesVc {
+        todo!()
+    }
+
+    #[turbo_tasks::function]
+    async fn ident(&self) -> Result<AssetIdentVc> {
+        let inner = self.full_module.ident().await?;
+
+        let query = StringVc::cell(format!("part={}", self.chunk_id).into());
+        Ok(AssetIdentVc::new(Value::new(AssetIdent {
+            query: Some(query),
+            ..(*inner).clone()
+        })))
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl EcmascriptChunkPlaceable for EcmascriptModulePartAsset {
+    #[turbo_tasks::function]
+    async fn as_chunk_item(
+        self_vc: EcmascriptModulePartAssetVc,
+        context: ChunkingContextVc,
+    ) -> Result<EcmascriptChunkItemVc> {
+        let s = self_vc.await?;
+
+        Ok(EcmascriptModulePartChunkItem {
+            module: self_vc,
+            context,
+            chunk_id: s.chunk_id,
+            full_module: s.full_module,
+            split_data: s.split_data,
+        }
+        .cell()
+        .into())
+    }
+
+    #[turbo_tasks::function]
+    async fn get_exports(self_vc: EcmascriptModuleAssetVc) -> Result<EcmascriptExportsVc> {
+        Ok(self_vc.analyze().await?.exports)
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl ChunkableAsset for EcmascriptModulePartAsset {
+    #[turbo_tasks::function]
+    async fn as_chunk(self_vc: EcmascriptModulePartAssetVc, context: ChunkingContextVc) -> ChunkVc {
+        EcmascriptChunkVc::new(context, self_vc.as_ecmascript_chunk_placeable()).into()
+    }
+}
+
+#[turbo_tasks::value]
+pub struct EcmascriptModulePartChunkItem {
+    full_module: EcmascriptModuleAssetVc,
+
+    split_data: SplitResultVc,
+
+    module: EcmascriptModulePartAssetVc,
+    context: ChunkingContextVc,
+
+    chunk_id: u32,
+}
+
+#[turbo_tasks::value_impl]
+impl EcmascriptModulePartAssetVc {
+    #[turbo_tasks::function]
+    async fn analyze(self) -> Result<AnalyzeEcmascriptModuleResultVc> {
+        let part = self.await?;
+        let this = part.full_module.await?;
+        Ok(analyze_ecmascript_module(
+            this.source,
+            part.full_module.as_resolve_origin(),
+            Value::new(this.ty),
+            this.transforms,
+            this.compile_time_info,
+            Some(part.chunk_id),
+        ))
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl ValueToString for EcmascriptModulePartChunkItem {
+    #[turbo_tasks::function]
+    async fn to_string(&self) -> Result<StringVc> {
+        Ok(StringVc::cell(format!(
+            "{} (ecmascript) -> part {}",
+            self.full_module.await?.source.ident().to_string().await?,
+            self.chunk_id
+        )))
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl EcmascriptChunkItem for EcmascriptModulePartChunkItem {
+    #[turbo_tasks::function]
+    async fn content(&self) -> Result<EcmascriptChunkItemContentVc> {
```

Comment:
The whole function looks very similar to `EcmascriptModuleChunkItem::content()`. Could we either share that code, or could we merge the whole struct into `EcmascriptModuleChunkItem` with `part_id: Option<u32>`